Gems/Terrain/Code/Source/TerrainRaycast/TerrainRaycastContext.cpp

```diff
@@ -0,0 +1,393 @@
+        // intersects in order from nearest to furthest until we find
+        // an intersection with the terrain or the ray exits the grid.
+        result = AzFramework::RenderGeometry::RayResult();
+        while (currentVoxel.GetMin().GetX() <= voxelGridMaxX &&
+               currentVoxel.GetMax().GetX() >= voxelGridMinX &&
+               currentVoxel.GetMin().GetY() <= voxelGridMaxY &&
+               currentVoxel.GetMax().GetY() >= voxelGridMinY &&
+               currentVoxel.GetMin().GetZ() <= voxelGridMaxZ &&
+               currentVoxel.GetMax().GetZ() >= voxelGridMinZ)
+        {
+            TriangulateAndFindNearestIntersection(terrainSystem,
+                                                  currentVoxel,
+                                                  rayStart,
+                                                  rayDirection,
+                                                  rayDirectionReciprocal,
+                                                  result);
+            if (result)
+            {
+                // Intersection found.
+                break;
+            }
+
+            // Step to the next voxel.
+            if (tMaxX < tMaxY && tMaxX < tMaxZ)
+            {
+                currentVoxelMin += stepVecX;
+                currentVoxelMax += stepVecX;
+                tMaxX += tDeltaX;
+            }
+            else if (tMaxY < tMaxZ)
+            {
+                currentVoxelMin += stepVecY;
+                currentVoxelMax += stepVecY;
+                tMaxY += tDeltaY;
+            }
+            else
+            {
+                currentVoxelMin += stepVecZ;
+                currentVoxelMax += stepVecZ;
+                tMaxZ += tDeltaZ;
+            }
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+TerrainRaycastContext::TerrainRaycastContext(TerrainSystem& terrainSystem)
+    : m_terrainSystem(terrainSystem)
+    , m_entityContextId(AzFramework::EntityContextId::CreateRandom())
+{
+    AzFramework::RenderGeometry::IntersectorBus::Handler::BusConnect(m_entityContextId);
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+TerrainRaycastContext::~TerrainRaycastContext()
+{
+    AzFramework::RenderGeometry::IntersectorBus::Handler::BusDisconnect();
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+AzFramework::RenderGeometry::RayResult TerrainRaycastContext::RayIntersect(
+    const AzFramework::RenderGeometry::RayRequest& ray)
+{
+    const AZ::Aabb terrainWorldBounds = m_terrainSystem.GetTerrainAabb();
+    const AZ::Vector2 terrainResolution = m_terrainSystem.GetTerrainHeightQueryResolution();
+    const AZ::Vector3 rayDirection = ray.m_endWorldPosition - ray.m_startWorldPosition;
+    const AZ::Vector3 rayDirectionReciprocal = rayDirection.GetReciprocal();
```

Comment:
Shouldn't the internal APIs you've created here take in the endWorldPosition as well, instead of just a direction?  Seems like there's a potential to end the iteration much earlier if the calling methods are only trying to project N meters in a direction.

Reply:
Good call, updated